### ISSUE TITLE
Directly pool.getConnection how automatic recovery connection resources

### ISSUE THREAD

**Issue by anhuisunfei** [felixge/node-mysql#1204]
``` javascript
var pool = mysql.createPool({
	host: '127.0.0.1',
	port: 3306,
	user: 'root',
	password: 'root' 
});
pool.getConnection(function(err, conn) {
    conn.query('select 1=1',function(){
    conn.relase();
  });
});
```

**Comment by anhuisunfei** [comment]
I want to automatically release connection pool

**Comment by anhuisunfei** [comment]
@dougwilson  I want to automatically release connection pool

**Comment by dougwilson** [comment]
Hi! I'm sorry, but I still don't understand the question :(

**Comment by anhuisunfei** [comment]
thank you @dougwilson

**Comment by dougwilson** [comment]
Perhaps you are looking for the following?
```js


var pool = mysql.createPool({
    host: '127.0.0.1',
    port: 3306,
    user: 'root',
    password: 'root' 
});
pool.query('select 1=1',function(){
   // you don't need conn.release() now
});
```

**Comment by anhuisunfei** [comment]
No, I just want to know there is no automatic release @dougwilson

**Comment by dougwilson** [comment]
If you call `getConnection`, you must always call `.release()` on the resulting connection; there is no way around that in an environment like Node.js (event loop). We provide you `pool.query` if you just need to do a single query without worrying about releasing.

**Comment by anhuisunfei** [comment]
Yes, I did some transaction processing。Ok, I see @dougwilson

**Comment by dougwilson** [comment]
If you only ever do a single transaction and want to automatically release on rollback or commit, you can do this yourself, though:
```js
var pool = mysql.createPool({
    host: '127.0.0.1',
    port: 3306,
    user: 'root',
    password: 'root' 
});

pool.getConnection(function(err, conn) {
  autoRelease(conn);

  conn.beginTransaction(function (err) {
    if (err) throw err;

    // ... do all your work in your transaction,
    // then calling conn.commit() or conn.rollback()
    // and your connection will automatically be released
    conn.query('INSERT ...');
    conn.query('UPDATE ...');
    conn.commit();
  });
});

function after(fn, cb) {
  return function () {
    fn.apply(this, arguments);
    cb();
  }
}

function autoRelease(conn) {
  conn.commit = after(conn.commit, release);
  conn.rollback = after(conn.rollback, release);

  function release() {
    conn.release();
  }
}
```

**Comment by anhuisunfei** [comment]
Good I try :+1:

**Comment by anhuisunfei** [comment]
``` javascript
var mysql = require('mysql');

var pool = mysql.createPool({
	host: '127.0.0.1',
	database:'shcem',
	port: 3306,
	user: 'root',
	password: 'root',
	debug: true
});



var execPool = function() {
	pool.getConnection(function(err, conn) {
		autoRelease(conn);

		conn.beginTransaction(function(err) {
			if (err) throw err;
			conn.query("INSERT INTO test(id,name,date,test) values(1,'123',now(),1)",
				function(err, ret) {
					if (err) {
						console.error(err);
						conn.rollback(function() {});
					} else {
						conn.query('UPDATE test set id=12321312 where id=23', function(err, ret) {
							if (err) {
								console.error(err);
								conn.rollback(function() {

								});
							} else {
								conn.commit(function() {
									console.log('success' + ret);
								});
							}
						});

					}
				});
		});
	});
}

function after(fn, cb) {
	return function() {
		fn.apply(this, arguments);
		cb();
	}
}

function autoRelease(conn) {
	conn.commit = after(conn.commit, release);
	conn.rollback = after(conn.rollback, release);

	function release() {
		if(conn)
		conn.release();
	}
}


setInterval(function(){
	execPool();
},1000);
```
After I tried to report this error===》
F:\MyCode\mysqlTest\node_modules\mysql\lib\protocol\Parser.js:82
        throw err;
              ^
Error: Connection already released
    at Pool.releaseConnection (F:\MyCode\mysqlTest\node_modules\mysql\lib\Pool.js:129:13)
    at PoolConnection.release (F:\MyCode\mysqlTest\node_modules\mysql\lib\PoolConnection.js:30:
    at release (F:\MyCode\mysqlTest\index.js:57:8)
    at PoolConnection.commit (F:\MyCode\mysqlTest\index.js:48:3)
    at Query._callback (F:\MyCode\mysqlTest\index.js:33:14)
    at Query.Sequence.end (F:\MyCode\mysqlTest\node_modules\mysql\lib\protocol\sequences\Sequen
    at Query._handleFinalResultPacket (F:\MyCode\mysqlTest\node_modules\mysql\lib\protocol\sequ
    at Query.OkPacket (F:\MyCode\mysqlTest\node_modules\mysql\lib\protocol\sequences\Query.js:7
    at Protocol._parsePacket (F:\MyCode\mysqlTest\node_modules\mysql\lib\protocol\Protocol.js:2
    at Parser.write (F:\MyCode\mysqlTest\node_modules\mysql\lib\protocol\Parser.js:77:12)

**Comment by sidorares** [comment]
That's because connection is monkey-patched twice. Change autoRelease to

```js
var Connection = require('mysql/lib/Connection.js');
function autoRelease(conn) {
    if (conn.commit != Connection.prototype.commit)
      conn.commit = after(conn.commit, release);
   if (conn.rollback != Connection.prototype.rollback) 
      conn.rollback = after(conn.rollback, release);

    function release() {
        if(conn)
        conn.release();
    }
}
```

**Comment by anhuisunfei** [comment]
@sidorares  not work!

**Comment by anhuisunfei** [comment]
``` javascript
var mysql = require('mysql'),
	Connection = require('mysql/lib/Connection.js');

var pool = mysql.createPool({
	host: '127.0.0.1',
	database: 'shcem',
	port: 3306,
	user: 'root',
	password: 'root',
	debug: false,
	connectionLimit: 3
});



var execPool = function() {
	pool.getConnection(function(err, conn) {
		autoRelease(conn);

		conn.beginTransaction(function(err) {
			if (err) throw err;
			conn.query("INSERT INTO test(id,name,date,test) values(1,'123',now(),1)",
				function(err, ret) {
					if (err) {
						console.error(err);
						conn.rollback(function() {});
					} else {
						console.log(ret);
						conn.query('UPDATE test set id=12321312 where id=1', function(err, ret) {
							if (err) {
								console.error(err);
								conn.rollback(function() {

								});
							} else {
								conn.commit(function() {
									console.log('success' +JSON.stringify(ret));
								});
							}
						});

					}
				});
		});
	});
}

function after(fn, cb) {
	console.log(12);
	return function() {
		fn.apply(this, arguments);
		cb();
	}
} 
function autoRelease(conn) {
	if (conn.commit != Connection.prototype.commit)
		conn.commit = after(conn.commit, release);
	if (conn.rollback != Connection.prototype.rollback)
		conn.rollback = after(conn.rollback, release);

	function release() {
		if (conn) { 
			conn.release();
		}
	} 
}


setInterval(function() {
	execPool();
}, 1000);
```

@sidorares

**Comment by anhuisunfei** [comment]
``` C#
static void Main(object[] args){
 
}
```

**Comment by sidorares** [comment]
comparison `if (conn.commit != Connection.prototype.commit)` should be actually `if (conn.commit == Connection.prototype.commit)`:

```js
var mysql = require('mysql'),
    Connection = require('mysql/lib/Connection.js');

var pool = mysql.createPool({
    host: '127.0.0.1',
    database: 'shcem',
    port: 3306,
    user: 'root',
    password: 'root',
    debug: false,
    connectionLimit: 3
});



var execPool = function() {
    pool.getConnection(function(err, conn) {
        autoRelease(conn);

        conn.beginTransaction(function(err) {
            if (err) throw err;
            conn.query("INSERT INTO test(id,name,date,test) values(1,'123',now(),1)",
                function(err, ret) {
                    if (err) {
                        console.error(err);
                        conn.rollback(function() {});
                    } else {
                        console.log(ret);
                        conn.query('UPDATE test set id=12321312 where id=1', function(err, ret) {
                            if (err) {
                                console.error(err);
                                conn.rollback(function() {

                                });
                            } else {
                                conn.commit(function() {
                                    console.log('success' +JSON.stringify(ret));
                                });
                            }
                        });

                    }
                });
        });
    });
}

function after(fn, cb) {
    console.log(12);
    return function() {
        fn.apply(this, arguments);
        cb();
    }
} 
function autoRelease(conn) {
    // if this is original conn.commit() and conn.rollback(), change them to be "commit and release" and "rollback and relese"
    if (conn.commit == Connection.prototype.commit) {
        conn.commit = after(conn.commit, release);
        conn.rollback = after(conn.rollback, release);
    }

    function release() {
        if (conn) { 
            conn.release();
        }
    } 
}


setInterval(function() {
    execPool();
}, 1000);
```

**Comment by anhuisunfei** [comment]
@sidorares :+1: is OK! Thank you very much

**Comment by anhuisunfei** [comment]
close

**Comment by anhuisunfei** [comment]
Closes #issuecomment-136938588

**Comment by anhuisunfei** [comment]
closesing @sidorares

**Comment by anhuisunfei** [comment]
This closes #1204

**Comment by anhuisunfei** [comment]
closes #issuecomment-136938588

**Comment by anhuisunfei** [comment]
closed

**Comment by anhuisunfei** [comment]
closed 1204#issuecomment-136938588

**Comment by anhuisunfei** [comment]
fixed #issuecomment-136938588

**Comment by anhuisunfei** [comment]
close #1206

**Comment by anhuisunfei** [comment]
closed #1206

**Comment by sidorares** [comment]
@anhuisunfei the issue already closed, no need to do anything
